docs/topics/downloader-middleware.rst

```diff
@@ -922,13 +922,72 @@ RobotsTxtMiddleware
     To make sure Scrapy respects robots.txt make sure the middleware is enabled
     and the :setting:`ROBOTSTXT_OBEY` setting is enabled.
 
+    This middleware has to be combined with a `robots.txt <http://www.robotstxt.org/>`_ parser.
+
+    Scrapy ships with support for three `robots.txt <http://www.robotstxt.org/>`_ parsers:
+
+    * `RobotFileParser`_
+    * `Reppy`_
+    * `Robotexclusionrulesparser`_
+
+    You can change the `robots.txt <http://www.robotstxt.org/>`_ parser with the :setting:`ROBOTSTXT_PARSER`
+    setting. Or you can also :ref:`implement support for a new parser <support-for-new-robots-parser>`.
+
 .. reqmeta:: dont_obey_robotstxt
 
 If :attr:`Request.meta <scrapy.http.Request.meta>` has
 ``dont_obey_robotstxt`` key set to True
 the request will be ignored by this middleware even if
 :setting:`ROBOTSTXT_OBEY` is enabled.
 
+.. _support-for-new-robots-parser:
+
+Implementing support for a new parser
+-------------------------------------
+
+You can implement support for a new `robots.txt <http://www.robotstxt.org/>`_ parser by implementing the
+interface described below.
+
+.. module:: scrapy.extensions.robotstxtparser
+
+.. class:: RobotsParser
+
+    .. method:: __init__(content)
+
+      Parse content of `robots.txt <http://www.robotstxt.org/>`_ file.
+
+      :param content: Content of `robots.txt <http://www.robotstxt.org/>`_ file. For Python 2.7, if ``content``
+                    contains non-ASCII characters, it should be ``UTF-8`` encoded. 
+      :type content: ``str``
+
+    .. method:: allowed(url, useragent)
+
+      Return ``True`` if ``url`` is allowed for crawling by the given ``useragent``, otherwise return ``False``.
+
+      :param url: URL to find crawling permission for. For Python 2.7, if ``url`` contains non-ASCII characters, 
```

Comment:
The parameter and return types should be ‘the same’ across Python versions (as in string vs bytes, Python 2 and Python 3 string types are not really the same underneath).

On a different note: What will the URL parameter that comes from Scrapy contain? Will it be the absolute, normalized URL? Just an absolute URL? Maybe not be an URL at all, and instead be an absolute path relative to the underlying domain (e.g. `/index.html` for `https://example.com/index.html`)?

Reply:
It will be an absolute URL but not normalised.

Reply:
> The parameter and return types should be ‘the same’ across Python versions (as in string vs bytes, Python 2 and Python 3 string types are not really the same underneath).

Could you elaborate more on this? Not sure, If I understood you completely here.

Reply:
Yes, my phrasing was very unfortunate.

The indicated type is `str`. However, you mention that in Python 2.7, if non-ASCII characters are used, UTF-8 encoding must be used.

The thing is, `str` values do not need an encoding at all, they are the decoded version of byte strings.

So, if the type is `str`, the sentence about encoding should be removed, it does not really apply. If the type were `bytes`, then the sentence should stay, but it should not be about Python 2.7 specifically, it would also apply to Python 3.